File: go/backtest/backtest.go
package backtest

import (
	"fmt"
	"math"

	"github.com/ducnhd/lotusmarket/go/technical"
	"github.com/ducnhd/lotusmarket/go/types"
)

type Config struct {
	Strategy string  `json:"strategy"` // "rsi", "ma_cross", "combined"
	Capital  float64 `json:"capital"`
	RSIBuy   float64 `json:"rsi_buy"`
	RSISell  float64 `json:"rsi_sell"`
}

type Trade struct {
	Date   string  `json:"date"`
	Action string  `json:"action"`
	Price  float64 `json:"price"`
	Shares float64 `json:"shares"`
	Value  float64 `json:"value"`
	Cash   float64 `json:"cash"`
	PnL    float64 `json:"pnl"`
}

type Result struct {
	Config        Config  `json:"config"`
	Trades        []Trade `json:"trades"`
	TotalReturn   float64 `json:"total_return_pct"`
	MaxDrawdown   float64 `json:"max_drawdown_pct"`
	WinRate       float64 `json:"win_rate_pct"`
	BuyHoldReturn float64 `json:"buy_hold_return_pct"`
	FinalCapital  float64 `json:"final_capital"`
	TradeCount    int     `json:"trade_count"`
}

// Run executes a backtest strategy on the provided price history.
// history must have at least 20 entries with Close prices.
func Run(cfg Config, history []types.StockData) (*Result, error) {
	if cfg.Capital <= 0 {
		cfg.Capital = 100_000_000 // 100M VND
	}
	if cfg.RSIBuy <= 0 {
		cfg.RSIBuy = 30
	}
	if cfg.RSISell <= 0 {
		cfg.RSISell = 70
	}
	if len(history) < 20 {
		return nil, fmt.Errorf("insufficient data: need 20+ days, got %d", len(history))
	}

	result := &Result{Config: cfg}

	closes := make([]float64, len(history))
	for i, h := range history {
		closes[i] = h.Close
	}

	cash := cfg.Capital
	shares := 0.0
	var trades []Trade
	peakValue := cfg.Capital
	maxDrawdown := 0.0
	wins := 0
	losses := 0

	startIdx := 50
	if startIdx >= len(history) {
		startIdx = 20
	}

	for i := startIdx; i < len(history); i++ {
		price := history[i].Close
		date := history[i].Date

		rsi := technical.RSI(closes[:i+1], 14)
		ma := technical.CalculateMovingAverages(closes[:i+1])

		buySignal := false
		sellSignal := false

		switch cfg.Strategy {
		case "rsi":
			buySignal = rsi < cfg.RSIBuy
			sellSignal = rsi > cfg.RSISell
		case "ma_cross":
			if ma.MA20 != nil && ma.MA50 != nil {
				buySignal = *ma.MA20 > *ma.MA50 && price > *ma.MA20
				sellSignal = *ma.MA20 < *ma.MA50 && price < *ma.MA20
			}
		case "combined":
			maBuy := ma.MA20 != nil && ma.MA50 != nil && *ma.MA20 > *ma.MA50
			maSell := ma.MA20 != nil && ma.MA50 != nil && *ma.MA20 < *ma.MA50
			buySignal = rsi < cfg.RSIBuy && maBuy
			sellSignal = rsi > cfg.RSISell || maSell
		default:
			buySignal = rsi < cfg.RSIBuy
			sellSignal = rsi > cfg.RSISell
		}

		if buySignal && shares == 0 && cash > 0 {
			shares = math.Floor(cash / price)
			if shares > 0 {
				cost := shares * price
				cash -= cost
				trades = append(trades, Trade{
					Date: date, Action: "BUY", Price: price,
					Shares: shares, Value: cost, Cash: cash,
				})
			}
		} else if sellSignal && shares > 0 {
			value := shares * price
			pnl := value - (shares * trades[len(trades)-1].Price)
			if pnl > 0 {
				wins++
			} else {
				losses++
			}
			cash += value
			trades = append(trades, Trade{
				Date: date, Action: "SELL", Price: price,
				Shares: shares, Value: value, Cash: cash, PnL: pnl,
			})
			shares = 0
		}

		totalValue := cash + shares*price
		if totalValue > peakValue {
			peakValue = totalValue
		}
		dd := (peakValue - totalValue) / peakValue * 100
		if dd > maxDrawdown {
			maxDrawdown = dd
		}
	}

	lastPrice := history[len(history)-1].Close
	result.FinalCapital = cash + shares*lastPrice
	result.TotalReturn = (result.FinalCapital - cfg.Capital) / cfg.Capital * 100
	result.MaxDrawdown = maxDrawdown
	result.Trades = trades
	result.TradeCount = len(trades)

	totalTrades := wins + losses
	if totalTrades > 0 {
		result.WinRate = float64(wins) / float64(totalTrades) * 100
	}

	firstPrice := history[startIdx].Close
	result.BuyHoldReturn = (lastPrice - firstPrice) / firstPrice * 100

	return result, nil
}

File: python/lotusmarket/backtest.py
"""Strategy backtesting engine for Vietnamese stocks."""

import math
from dataclasses import dataclass, field
from typing import List
import pandas as pd
from lotusmarket.technical import rsi_value, ma_value


@dataclass
class BacktestConfig:
    strategy: str = "rsi"  # "rsi", "ma_cross", "combined"
    capital: float = 100_000_000  # 100M VND
    rsi_buy: float = 30.0
    rsi_sell: float = 70.0


@dataclass
class BacktestTrade:
    date: str = ""
    action: str = ""
    price: float = 0.0
    shares: float = 0.0
    value: float = 0.0
    cash: float = 0.0
    pnl: float = 0.0


@dataclass
class BacktestResult:
    config: BacktestConfig = field(default_factory=BacktestConfig)
    trades: List[BacktestTrade] = field(default_factory=list)
    total_return: float = 0.0
    max_drawdown: float = 0.0
    win_rate: float = 0.0
    buy_hold_return: float = 0.0
    final_capital: float = 0.0
    trade_count: int = 0


def run(cfg: BacktestConfig, prices: pd.Series) -> BacktestResult:
    """Run a backtest strategy on price series.

    Args:
        cfg: Backtest configuration
        prices: pandas Series of close prices (at least 20 values)

    Returns:
        BacktestResult with trades, returns, drawdown, win rate
    """
    if len(prices) < 20:
        raise ValueError(f"Insufficient data: need 20+ prices, got {len(prices)}")

    if cfg.capital <= 0:
        cfg.capital = 100_000_000
    if cfg.rsi_buy <= 0:
        cfg.rsi_buy = 30
    if cfg.rsi_sell <= 0:
        cfg.rsi_sell = 70

    result = BacktestResult(config=cfg)
    closes = prices.values.tolist()

    cash = cfg.capital
    shares = 0.0
    trades = []
    peak_value = cfg.capital
    max_dd = 0.0
    wins = 0
    losses = 0

    start_idx = min(50, len(closes) - 1)
    if start_idx < 20:
        start_idx = 20

    for i in range(start_idx, len(closes)):
        price = closes[i]
        sub = pd.Series(closes[: i + 1])

        r = rsi_value(sub, 14)
        ma20 = ma_value(sub, 20)
        ma50 = ma_value(sub, 50)

        buy_signal = False
        sell_signal = False

        if cfg.strategy == "rsi":
            buy_signal = r < cfg.rsi_buy
            sell_signal = r > cfg.rsi_sell
        elif cfg.strategy == "ma_cross":
            if ma20 is not None and ma50 is not None:
                buy_signal = ma20 > ma50 and price > ma20
                sell_signal = ma20 < ma50 and price < ma20
        elif cfg.strategy == "combined":
            ma_buy = ma20 is not None and ma50 is not None and ma20 > ma50
            ma_sell = ma20 is not None and ma50 is not None and ma20 < ma50
            buy_signal = r < cfg.rsi_buy and ma_buy
            sell_signal = r > cfg.rsi_sell or ma_sell
        else:
            buy_signal = r < cfg.rsi_buy
            sell_signal = r > cfg.rsi_sell

        if buy_signal and shares == 0 and cash > 0:
            shares = math.floor(cash / price)
            if shares > 0:
                cost = shares * price
                cash -= cost
                trades.append(
                    BacktestTrade(
                        date=str(i),
                        action="BUY",
                        price=price,
                        shares=shares,
                        value=cost,
                        cash=cash,
                    )
                )
        elif sell_signal and shares > 0:
            value = shares * price
            pnl = value - (shares * trades[-1].price)
            if pnl > 0:
                wins += 1
            else:
                losses += 1
            cash += value
            trades.append(
                BacktestTrade(
                    date=str(i),
                    action="SELL",
                    price=price,
                    shares=shares,
                    value=value,
                    cash=cash,
                    pnl=pnl,
                )
            )
            shares = 0

        total_value = cash + shares * price
        if total_value > peak_value:
            peak_value = total_value
        dd = (peak_value - total_value) / peak_value * 100
        if dd > max_dd:
            max_dd = dd

    last_price = closes[-1]
    result.final_capital = cash + shares * last_price
    result.total_return = (result.final_capital - cfg.capital) / cfg.capital * 100
    result.max_drawdown = max_dd
    result.trades = trades
    result.trade_count = len(trades)

    total_trades = wins + losses
    if total_trades > 0:
        result.win_rate = wins / total_trades * 100

    first_price = closes[start_idx]
    result.buy_hold_return = (last_price - first_price) / first_price * 100

    return result
